### PR TITLE
Analyze code coverage

### PR DESCRIPTION
**Pull request by nulltoken**
- Should be triggered ~~as part of the scheduled build~~ upon the merge of a PR
- See https://github.com/csMACnz/Coveralls.net-Samples/tree/xunit-opencover-appveyor for inspiration

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -98,7 +105,18 @@ test_script:
 - ps: |
     If ($Env:SHOULD_RUN_COVERITY_ANALYSIS -eq $False)
     {
-      & "$Env:xunit_runner" "$Env:APPVEYOR_BUILD_FOLDER\LibGit2Sharp.Tests\bin\Release\LibGit2Sharp.Tests.dll" /appveyor
+      If ($Env:SHOULD_PACKAGE_NUGET_ARTIFACT -eq $True -and $Env:publish_on_success -eq $True)
```
Comment (dahlbyk) [inline appveyor.yml]: Not sure why coverage depends on `$Env:SHOULD_PACKAGE_NUGET_ARTIFACT`?

Comment (nulltoken) [inline appveyor.yml]: You're right. The idea is to analyze on merges. This also happen to be when we generate NuGet packages.

I think the variable is badly named. What would you call it?

Comment (dahlbyk) [inline appveyor.yml]: Personally I'd add a separate `SHOULD_RUN_COVERALLS` variable.

File: appveyor.yml [hunk 2]
```diff
@@ -120,6 +138,18 @@ after_test:
...
+
+      .\packages\coveralls.net\csmacnz.Coveralls.exe `
+        --opencover -i opencoverCoverage.xml `
+        --repoToken $env:COVERALLS_REPO_TOKEN `
+        --commitId $env:APPVEYOR_REPO_COMMIT `
+        --commitBranch $env:APPVEYOR_REPO_BRANCH `
+        --commitAuthor $env:APPVEYOR_REPO_COMMIT_AUTHOR `
+        --commitEmail $env:APPVEYOR_REPO_COMMIT_AUTHOR_EMAIL `
+        --commitMessage $env:APPVEYOR_REPO_COMMIT_MESSAGE `
+        --jobId $env:APPVEYOR_JOB_ID
```
Comment (nulltoken) [inline appveyor.yml]: Hmmm. **[Build log](https://ci.appveyor.com/project/libgit2/libgit2sharp/build/1002/job/qksc4c9c4oxxwlqi#L339)** shows the following

```
Usage:
  csmacnz.Coveralls (--opencover | --dynamiccodecoverage | --monocov) -i ./opencovertests.xml --repoToken repoToken> [-o ./opencovertests.json] [--dryrun] [--useRelativePaths [--basePath <path>] ] [--commitId commitId> --commitBranch <commitBranch> [--commitAuthor <commitAuthor> --commitEmail <commitEmail> --commitMessage <commitMessage>] ] [--jobId <jobId>] [--serviceName <Name>]
  csmacnz.Coveralls --version
  csmacnz.Coveralls --help
```

I may have messed up something, but I can't find what. The log doesn't show any of the troubleshooting messages than can be found in the **[source code](https://github.com/csMACnz/coveralls.net/blob/master/src/csmacnz.Coveralls/Program.cs)**

@csMACnz I wrote this based on https://github.com/csMACnz/Coveralls.net-Samples/tree/xunit-opencover-appveyor Any idea? Help? :pray:

**Comment by csMACnz** [comment]
The problem is that appveyor removes secure token variables on PR builds. The args list becomes 
    ... --repoToken  --commitId $env:...
Which is invalid syntax. I'm still looking at how to solve this usage error ao it becomes more obvious.

If you look at my other projects, I wrap it in a condition to check for the PR build environment variable

File: appveyor.yml (hunk 1 above)
Comment (nulltoken) [inline appveyor.yml]: Done.

**Comment by nulltoken** [comment]
@csMACnz Thanks for your answer.

I've pushed this as a separate branch (see https://github.com/libgit2/libgit2sharp/compare/ntk/coverage...coveralls).

The **[log](https://ci.appveyor.com/project/libgit2/libgit2sharp/build/1019/job/m8c087jh1mgafr26)** still displays the same usage pattern.

However, this time  it also fails with a `Command exited with code 1` message, still without any troubleshooting message. Could it be possible that the `Console.Error.WriteLine()` calls aren't intercepted by AppVeyor?

Another question. When the Usage is displayed, the **[possible options](https://github.com/csMACnz/coveralls.net/blob/45780f7aa8a7dfcfc36b4419963aed3dafd0189b/src/csmacnz.Coveralls/T4DocoptNet.cs#L16-L34)** aren't.  Is that expected?

/cc @feodorfitsner

**Comment by FeodorFitsner** [comment]
You mean AppVeyor swallows `Console.Error.WriteLine()` from .NET console app?

**Comment by nulltoken** [comment]
@FeodorFitsner It *might*. By taking a look at **[source code](https://github.com/csMACnz/coveralls.net/blob/master/src/csmacnz.Coveralls/Program.cs)** it *looks* like at least one of them should be output in this **[log](https://ci.appveyor.com/project/libgit2/libgit2sharp/build/1019/job/m8c087jh1mgafr26)**.

It's fairly possible I may be completely wrong about this though.

**Comment by csMACnz** [comment]
Your variables don't match so its still the same problem.  You declare coveralls_token then use COVERALLS_REPO_TOKEN. 
Hopefully making them the same fixes your problem.

Also, the docopts.net library I use is printing the usage statements, and returning the exit code. That is why none of my console.error.writeline calls show up.

**Comment by nulltoken** [comment]
>Your variables don't match so its still the same problem. You declare coveralls_token then use COVERALLS_REPO_TOKEN. 

@csMACnz :gem: Thanks a lot for having pointed me at this mistake! 

@dahlbyk @jamill The `coveralls` branch code coverage has been successfully pushed up to `coveralls.io`. One can peek at the published analysis at https://coveralls.io/r/libgit2/libgit2sharp

I'm going to update the `appveyor.yml` file with some Console output in order to ease the troubleshooting of the build steps being run. I'll ping you when it's done.

@csMACnz Amazingly neat package and **AWESOME** support :bangbang:

**Comment by nulltoken** [comment]
**TODO:**

 - Maybe not consider files under `LibGit2Sharp.Tests` namespace for coverage analysis (see **[Constants.cs](https://coveralls.io/builds/2868860/source?filename=c%3A%2Fprojects%2Flibgit2sharp%2FLibGit2Sharp.Tests%2FTestHelpers%2FConstants.cs)**)
- Correctly configure the project root path (`c:/projects/libgit2sharp/`)

**Comment by dahlbyk** [comment]
> Maybe not consider files under LibGit2Sharp.Tests namespace for coverage analysis (see Constants.cs)

I see you've already done this (cb5d495e5518d2b52bc938869c3d8a087f4e47c7) but :+1: anyway.

> Correctly configure the project root path (`c:/projects/libgit2sharp/`)

This looks correct now as well.

Coverage numbers look correct. Low coverage on exceptions makes it kind of hard to identify actual gaps in coverage. I noticed most of our "for mocking purposes" ctors were untested, so just pushed a commit to `coveralls` to see how many "false negatives" we can avoid by actually trying to create mock instances.

**Comment by nulltoken** [comment]
>I noticed most of our "for mocking purposes" ctors were untested, so just pushed a commit to `coveralls` to see how many "false negatives" we can avoid by actually trying to create mock instances.

:heart_eyes: It's just been published!

**Comment by nulltoken** [comment]
@dahlbyk As this looks like working, I'm going to rewrite now both this PR and the `coveralls` branch. I'll ping you when done.

**Comment by nulltoken** [comment]
@dahlbyk Done. Let's wait for the report to be published to coveralls.io. Once it's available, how about having this one merged?

**Comment by dahlbyk** [comment]
:shipit:

**Comment by nulltoken** [comment]
@dahlbyk Thanks for the help and the feedback :heart: 

I've dropped all the coverage analysis. One should be automatically run during the next scheduled build (for now, daily at 00:00UTC)

**Pull request merged**